keps/sig-api-machinery/1623-standardize-conditions/README.md

```diff
@@ -0,0 +1,162 @@
+
+## Proposal
+
+Introduce a type into k8s.io/apimachinery/pkg/apis/meta/v1 for `Condition` that looks like 
+```go
+type Condition struct {
+	// Type of condition in CamelCase or in foo.example.com/CamelCase.
+	// Many .condition.type values are consistent across resources like Available, but because arbitrary conditions can be
+	// useful (see .node.status.conditions), the ability to deconflict is important.
+	// +required
+	Type string `json:"type" protobuf:"bytes,1,opt,name=type"`
+	// Status of the condition, one of True, False, Unknown.
+	// +required
+	Status ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status"`
+	// If set, this represents the .metadata.generation that the condition was set based upon.
+	// For instance, if .metadata.generation is currently 12, but the .status.condition[x].observedGeneration is 9, the condition is out of date
+	// with respect to the current state of the instance.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty" protobuf:"varint,3,opt,name=observedGeneration"`
+	// Last time the condition transitioned from one status to another.
+	// This should be when the underlying condition changed.  If that is not known, then using the time when the API field changed is acceptable.
+	// +required
+	LastTransitionTime metav1.Time `json:"lastTransitionTime" protobuf:"bytes,4,opt,name=lastTransitionTime"`
+	// The reason for the condition's last transition in CamelCase.
+	// The specific API may choose whether or not this field is considered a guaranteed API.
+	// This field may not be empty.
+	// +required
+	Reason string `json:"reason" protobuf:"bytes,5,opt,name=reason"`
+	// A human readable message indicating details about the transition.
+	// This field may be empty.
+	// +required
+	Message string `json:"message" protobuf:"bytes,6,opt,name=message"`
+}
+```
+
+This is not strictly compatible with any of our existing conditions because of either proto ordinals,
+required vs optional, or omitEmpty or not.
+However, it encapsulates the best of what we've learned and will allow new APIs to have a unified type.
+
+### Noteworthy choices
+ 1. `lastTransitionTime` is required.
+    Some current implementations allow this to be missing, but this makes it difficult for consumers.
+    By requiring it, the actor setting the field can set it to the best possible value instead of having clients try to guess.
+ 2. `reason` is required and must not be empty.
+    The actor setting the value should always describe why the condition is the way it is, even if that value is "unknown unknowns".
+    No other actor has the information to make a better choice.
+ 3. `lastHeartbeatTime` is removed.
+    This field caused excessive write loads as we scaled.
+    If an API needs this concept, it should codify it separately and possibly using a different resource.
+ 4. `observedGeneration` does not follow the standard requirement of "all optional fields are pointers".
```

Comment:
Worked through the reasoning of an optional, non-pointer field in this case with @lavalamp and @liggitt .  This is where I think we landed.

Reply:
this makes sense to me

Reply:
If we never start committing to this, we'll never reach consistency.

I waive my concern - it's not worth arguing about piece by piece.  It doesn't matter much for JSON.  I would expect it changes the on-wire proto representation, but on inspection it seems not.  That is surprising to me, but the proto marshal/unmarshal code is human-hostile, so ... `<shrug>`

Reply:
The rule is, the zero value is sufficiently invalid that it doesn't create any ambiguity. Omitted-on-the-wire needs to mean "I have no opinion". As long as that can't be confused for "I have a positive opinion that it should be the zero value", it's fine.

I also am a little unhappy about this making things more confusing. Basically the argument is that it's worth the convenience of not having the pointer in go.